src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract partial class AbstractTypeImportCompletionProvider : CommonCompletionProvider
+    {
+        private readonly SyntaxAnnotation _annotation = new SyntaxAnnotation();
+        private readonly ITypeImportCompletionService _typeImportCompletionService;
+        private readonly IExperimentationService _experimentationService;
+
+        public AbstractTypeImportCompletionProvider(Workspace workspace)
+        {
+            _typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+            _experimentationService = workspace.Services.GetService<IExperimentationService>();
+        }
+
+        protected abstract Task<SyntaxContext> CreateContextAsync(Document document, int position, CancellationToken cancellationToken);
+
+        protected abstract ImmutableHashSet<string> GetNamespacesInScope(SyntaxNode location, SemanticModel semanticModel, CancellationToken cancellationToken);
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+            var document = completionContext.Document;
+            var cancellationToken = completionContext.CancellationToken;
+
+            var importCompletionOptionValue = completionContext.Options.GetOption(CompletionOptions.ShowImportCompletionItems, document.Project.Language);
+
+            // Don't trigger import completion if the option value is "default" and the experiment is disabled for the user. 
+            if (importCompletionOptionValue == false ||
+                (importCompletionOptionValue == null && _experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.TypeImportCompletion) != true))
+            {
+                return;
+            }
+
+            var syntaxContext = await CreateContextAsync(document, completionContext.Position, cancellationToken).ConfigureAwait(false);
+            await AddCompletionItemsAsync(document, syntaxContext, completionContext, cancellationToken).ConfigureAwait(false);
```

Comment:
nit, order syntaxContext after completionContext since it is 'narrower' in scope.  i.e. completion context is what is passed in/out of this method containing method, while syntaxContext is only passed in/out of the inner method.

Reply:
nit: don't need to pass Document along.

Reply:
nit: considre just having one single method. not sure i see the value in two here.